### PR TITLE
Asynchronous message handling

### PR DESCRIPTION
**Pull request by nikkolasg**
Finally !
The incoming network messages are now buffered in the node and then dispatched as soon as possible to the protocol instance of the node.
Some might say "why don't just use buffered channel", well, you might be right BUT:
* With this approach we don't allocate as more than we need, and we can actually control our allocation if let's say there was a burst at some point, allocated a huge amount of memory, then it went back to normal, we *could* reduce the amount of memory by just taking a smaller chunk of the buffer and let the GC doing its job.
* It's inspired from [here](http://marcio.io/2015/07/handling-1-million-requests-per-minute-with-golang/). I kinda like their approach and took a simplified version. But in the long run, for all the messages that we are going to have (external api + tree/list handling + *many* protocol instances etc), it might make sense to implement something like this at a more lower level, at host.go with many *workers* doing the dispatching. So I see it as a training!
* I had fun doing it even if there was one ***** stupid bug.

If at any point you feel this is **overkill**, please **say so**! I am also OK to use a buffered channel and checking the `len(channel)` each time a new message arrives. Actually I did not know about the fact that we can query the actual *length* of a buffered channel before the jogging of this afternoon... :)

### CODE REVIEW

**Comment by ineiti** [comment]
+1 for overkill ;) I have my own solution in #290 - a simple queueReader with a queueWriter does the same.

I copied your test, but then it passed even without me changing anything in the `node.go`... So I did change it a bit.

**Comment by Liamsi** [comment]
> +1 for overkill ;) I have my own solution in #290 - a simple queueReader with a queueWriter does the same.

I though you read in the bus ;-) @ineiti's solution looks indeed simpler (easier to understand -> less error prone).

File: lib/sda/node.go [hunk 1]
```diff
@@ -347,9 +359,15 @@ func (n *Node) DispatchMsg(sdaMsg *SDAData) error {
...
 	default:
-		return errors.New("This message-type is not handled by this protocol")
+		err = errors.New("This message-type is not handled by this protocol")
+	}
+	// XXX Should at some point define a central way / point to handle errors
+	// ... ?
+	// Before the error was just going up to the sda.Host and nothing else was
+	// taken into account.
+	if err != nil {
+		dbg.Error(err)
```
Comment (Liamsi) [inline lib/sda/node.go]: At least print it out (silently omitting it, can create boring, hence seemingly endless debugging sessions). Pass it up if you think there are reasons why the caller should handle it.

**Comment by Liamsi** [comment]
If you both can't agree on what approach is better: Let's benchmark both ;-)

**Comment by ineiti** [comment]
Whatever approach we take, there is a lot of work to do afterwards. I spent most of yesterday's evening debugging all concurrency issues now that we can have multiple messages waiting. Locks are now all over the place.

**Comment by Liamsi** [comment]
Let @nikkolasg fix the test. I'll write a benchmark for both and let's discuss and decide afterwards.

**Comment by nikkolasg** [comment]
No need to benchmark @Liamsi it's fine let's go with the simpler approach for the moment ;)
However I'll comment on #290 there's some deadlock issues.

**Comment by nikkolasg** [comment]
Droped in favor of #290

**Pull request closed without merging**